### ISSUE TITLE
Errors running with docker-compose

### ISSUE THREAD

**Issue by stmcginnis** [chaoss/grimoirelab#371]
I am following along with the instructions listed on:

https://github.com/chaoss/grimoirelab/blob/master/docker-compose/README.md

Still digging in, but looks like some of the services are hitting errors during startup that are resulting in a non-functional system.

Here are some system details:

```
$ lsb_release -d
Description:	Ubuntu 18.04.5 LTS
$ git --version
git version 2.17.1
$ docker --version
Docker version 19.03.6, build 369ce74a3c
$ docker-compose --version
docker-compose version 1.17.1, build unknown
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            3.9G     0  3.9G   0% /dev
tmpfs           798M  680K  797M   1% /run
/dev/vda1        39G  7.3G   32G  19% /
tmpfs           3.9G     0  3.9G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           3.9G     0  3.9G   0% /sys/fs/cgroup
/dev/vda15      105M  3.6M  101M   4% /boot/efi
tmpfs           798M     0  798M   0% /run/user/1000
```

The first error I see is in the hatstall output:

```
hatstall_1       | Running Hatstall
hatstall_1       | cp: cannot create regular file 'grimoirelab-hatstall/django-hatstall/shdb.cfg': Permission denied
hatstall_1       | Site 000-default already disabled
hatstall_1       | Site apache-hatstall already enabled
hatstall_1       | Considering dependency setenvif for ssl:
hatstall_1       | Module setenvif already enabled
hatstall_1       | Considering dependency mime for ssl:
hatstall_1       | Module mime already enabled
hatstall_1       | Considering dependency socache_shmcb for ssl:
hatstall_1       | Module socache_shmcb already enabled
hatstall_1       | Module ssl already enabled
hatstall_1       | sed: couldn't open temporary file /home/grimoirelab/grimoirelab-hatstall/django-hatstall/django_hatstall/sed6O7J6F: Permission denied
hatstall_1       | Traceback (most recent call last):
hatstall_1       |   File "./config_deployment.py", line 42, in <module>
hatstall_1       |     with open(settings_file, "w") as f:
hatstall_1       | PermissionError: [Errno 13] Permission denied: 'django_hatstall/settings.py'
hatstall_1       | No changes detected
```

There is a kibiter error (maybe a timing issue?):

```
kibiter_1        |  FATAL  Error: Re-index failed :: 2 of 5 shards failed. Check Elasticsearch cluster health for more information.
```

elasticsearch appears to still be initializing, but all I see after that is:

```
elasticsearch_1  | [2020-08-18T16:05:35,257][INFO ][o.e.c.r.a.AllocationService] [fwsss97] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[.kibana_2][0], [.kibana][2], [.kibana][3], [.kibana_1][0]] ...]).
```

Any suggestions where to dig deeper or other things to check?

**Comment by vchrombie** [comment]
Hi @stmcginnis, thanks for opening the issue and sorry for the inconvenience caused. Let me see if I can help you with this.

> The first error I see is in the hatstall output:
> 
> ```
> hatstall_1       | Running Hatstall
> hatstall_1       | cp: cannot create regular file 'grimoirelab-hatstall/django-hatstall/shdb.cfg': Permission denied
> hatstall_1       | Site 000-default already disabled
> hatstall_1       | Site apache-hatstall already enabled
> hatstall_1       | Considering dependency setenvif for ssl:
> hatstall_1       | Module setenvif already enabled
> hatstall_1       | Considering dependency mime for ssl:
> hatstall_1       | Module mime already enabled
> hatstall_1       | Considering dependency socache_shmcb for ssl:
> hatstall_1       | Module socache_shmcb already enabled
> hatstall_1       | Module ssl already enabled
> hatstall_1       | sed: couldn't open temporary file /home/grimoirelab/grimoirelab-hatstall/django-hatstall/django_hatstall/sed6O7J6F: Permission denied
> hatstall_1       | Traceback (most recent call last):
> hatstall_1       |   File "./config_deployment.py", line 42, in <module>
> hatstall_1       |     with open(settings_file, "w") as f:
> hatstall_1       | PermissionError: [Errno 13] Permission denied: 'django_hatstall/settings.py'
> hatstall_1       | No changes detected
> ```

I found an issue which is almost similar. https://github.com/chaoss/grimoirelab-hatstall/issues/110

According to @zhquan's answer, I think it will be resolved after providing the right permissions. 
>> Make sure the user `docker` has `read` and `write` permission on `grimoirelab-hatstall/django-hatstall/shdb.cfg` or set `777` if you don't know how to change it.

It would be great if you can share your `docker-compose.yml` file if you need more help.

> elasticsearch appears to still be initializing, but all I see after that is:
> 
> ```
> elasticsearch_1  | [2020-08-18T16:05:35,257][INFO ][o.e.c.r.a.AllocationService] [fwsss97] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[.kibana_2][0], [.kibana][2], [.kibana][3], [.kibana_1][0]] ...]).
> ```

I would suggest increasing the virtual memory for Elasticsearch if you haven't already.
```
$ sysctl -w vm.max_map_count=262144
```

Also, I think there are multiple indexes (saw `kibana_1`, `kibana_2`) being created. You can confirm this by checking the indices after the ES is started. You can resolve the problem by deleting the extra indices if there are any. Please share the output if you are not sure about it.

(let's say you are using the secured docker-compose solution) 
```
$ curl -XGET -k "https://admin:admin@localhost:9200/_cat/indices?v"
$ curl -XDELETE -k https://admin:admin@localhost:9200/index_you_wish_to_delete

```
Another way, after Kibiter is started, go to `localhost:5601` and check indices in the Dev Console (:wrench: in the left sidebar).
```
GET /_cat/indices
DELETE index_you_wish_to_delete
```

> There is a kibiter error (maybe a timing issue?):
> 
> ```
> kibiter_1        |  FATAL  Error: Re-index failed :: 2 of 5 shards failed. Check Elasticsearch cluster health for more information.
> ```

I think this will be fixed once the elasticsearch is working fine.

Let us know if you need more help. :slightly_smiling_face:

**Comment by stmcginnis** [comment]
Thanks @vchrombie! I am using the docker-compose file from the repo. My steps were just the following:

```
$ # install docker, git, etc.
$ sysctl -w vm.max_map_count=262144
$ git clone https://github.com/chaoss/grimoirelab
$ cd grimoirelab/docker-compose
$ docker-compose up
```

That was a good tip with the hatstall issue. I ran:

```
$ chmod 666 ../default-grimoirelab-settings/shdb.cfg
$ docker-compose up
```

That got farther. It appears hatstall and elasticsearch now start up successfully, so that was definitely one of the issues.

The next issue I see is with kibiter. It ends up emitting the following error (truncated since it's really long):

```
{"type":"log","@timestamp":"2020-08-19T15:42:37Z","tags":["fatal","root"],"pid":1,"message":"{ [invalid_alias_name_exception] Invalid alias name [.kibana], an index exists with the same name as the alias, with { index_uuid=\"RKF7v5vsRXqzyivqUSJtEw\" & index=\".kibana\" }...
```

And if I run the curl command you suggested, it does indeed show .kibana and others:

```
health status index     uuid                   pri rep docs.count docs.deleted store.size pri.store.size
green  open   .kibana_1 m1xKTdT0RV210bkTThltRQ   1   0          0            0       261b           261b
green  open   .kibana_2 dK4EYp4GQyygZRp5npARMA   1   0          0            0       261b           261b
yellow open   .kibana   FHQy2nNcRpKPiZbzq9VfQg   5   1          0            0      1.2kb          1.2kb
```

Things seem steady after that for a few minutes, then I start getting output from mordred:

```
mordred_1        | 2020-08-19 15:51:49,687 - sirmordred.task_manager - ERROR - [Global tasks] Exception in Task Manager HTTPConnectionPool(host='kibiter', port=5601): Max retries exceeded with url: /api/kibana/settings/defaultIndex (Caused by ResponseError('too many 503 error responses',))
mordred_1        | Traceback (most recent call last):
mordred_1        |   File "/usr/local/lib/python3.5/dist-packages/kidash/kidash.py", line 1431, in check_kibana_index
mordred_1        |     res.raise_for_status()
mordred_1        |   File "/usr/local/lib/python3.5/dist-packages/requests/models.py", line 940, in raise_for_status
mordred_1        |     raise HTTPError(http_error_msg, response=self)
mordred_1        | requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://elasticsearch:9200/.kibana
```

I am assuming that that one is a side effect of the kibiter issue though.

**Comment by vchrombie** [comment]
Hi @stmcginnis, thanks for explaining each your steps. 

> That got farther. It appears hatstall and elasticsearch now start up successfully, so that was definitely one of the issues.

That's great to hear. :slightly_smiling_face: 

> The next issue I see is with kibiter. It ends up emitting the following error (truncated since it's really long):
> 
> ```
> {"type":"log","@timestamp":"2020-08-19T15:42:37Z","tags":["fatal","root"],"pid":1,"message":"{ [invalid_alias_name_exception] Invalid alias name [.kibana], an index exists with the same name as the alias, with { index_uuid=\"RKF7v5vsRXqzyivqUSJtEw\" & index=\".kibana\" }...
> ```

I was expecting this problem `an index exists with the same name as the alias`.

>> Also, I think there are multiple indexes (saw kibana_1, kibana_2) being created.

> And if I run the curl command you suggested, it does indeed show .kibana and others:
> 
> ```
> health status index     uuid                   pri rep docs.count docs.deleted store.size pri.store.size
> green  open   .kibana_1 m1xKTdT0RV210bkTThltRQ   1   0          0            0       261b           261b
> green  open   .kibana_2 dK4EYp4GQyygZRp5npARMA   1   0          0            0       261b           261b
> yellow open   .kibana   FHQy2nNcRpKPiZbzq9VfQg   5   1          0            0      1.2kb          1.2kb
> ```

Yes, now it is kinda clear. As you can see there are multiple indices being created like `kibana_1` and `kibana_2`. It won't be smooth if there are multiple indices. So, the quick fix is to delete the other index which is not necessary.

When you run the below command, it will provide you all aliases.
```
$ curl -XGET -k "https://admin:admin@localhost:9200/_alias?pretty"
```

The correct configuration output would look something like
```
{
  ".kibana_1" : {
    "aliases" : {
      ".kibana" : { }
    }
  },
  ...
  ...
}
```

I guess your configuration will be something like (this is an assumption :grimacing:)
```
{
  ".kibana_1" : {
    "aliases" : {
      ".kibana" : { }
    }
  },
  ".kibana_2" : {
  },
  ...
  ...
}
```

So, you can delete the `kibana_2`, according to the example I have given.
```
$ curl -XDELETE -k https://admin:admin@localhost:9200/.kibana_2
```

It would be better if you can restart the docker-compose, and I think it should fix this. If you need more help, please share the aliases too. 

> Things seem steady after that for a few minutes, then I start getting output from mordred:
> 
> ```
> mordred_1        | 2020-08-19 15:51:49,687 - sirmordred.task_manager - ERROR - [Global tasks] Exception in Task Manager HTTPConnectionPool(host='kibiter', port=5601): Max retries exceeded with url: /api/kibana/settings/defaultIndex (Caused by ResponseError('too many 503 error responses',))
> mordred_1        | Traceback (most recent call last):
> mordred_1        |   File "/usr/local/lib/python3.5/dist-packages/kidash/kidash.py", line 1431, in check_kibana_index
> mordred_1        |     res.raise_for_status()
> mordred_1        |   File "/usr/local/lib/python3.5/dist-packages/requests/models.py", line 940, in raise_for_status
> mordred_1        |     raise HTTPError(http_error_msg, response=self)
> mordred_1        | requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://elasticsearch:9200/.kibana
> ```
> 
> I am assuming that that one is a side effect of the kibiter issue though.

Yes, I think this won't show up if once we fix the Kibiter issue.

Let us know how it goes. :slightly_smiling_face:

**Comment by stmcginnis** [comment]
> It would be better if you can restart the docker-compose, and I think it should fix this. If you need more help, please share the aliases too.

This has happened multiple times after cleaning up and re-running docker-compose. These appear to be created by the start up process, so I can delete them, but it would be good to figure out where this is coming from.

Are you saying this process works fine for you? Can you provide some system information so I can see what might be different?

**Comment by stmcginnis** [comment]
Not sure if this helps, but I kept fetching those aliases during start up. It first created .kibana_1, then .kibana, then .kibana_2. Final result is:

```
$ curl -XGET -k "http://admin:admin@localhost:9200/_alias?pretty"
{
  ".kibana" : {
    "aliases" : { }
  },
  ".kibana_1" : {
    "aliases" : { }
  },
  ".kibana_2" : {
    "aliases" : { }
  }
}
```

**Comment by vchrombie** [comment]
> > It would be better if you can restart the docker-compose, and I think it should fix this. If you need more help, please share the aliases too.
> 
> This has happened multiple times after cleaning up and re-running docker-compose. 

Sorry for not being clear. After you delete the extra index, you can restart the docker-compose `docker-compose up`.

> These appear to be created by the start up process, so I can delete them, but it would be good to figure out where this is coming from.

I have faced it a couple of times before I cannot remember if it was during the first time

> Are you saying this process works fine for you? Can you provide some system information so I can see what might be different?

Yes, this works fine for me. I personally use the docker-compose solution for elasticsearch and kibiter. So, I hope it should work fine for you as well.

**Comment by stmcginnis** [comment]
Full output of the docker-compose up initialization can be found here:

https://pastebin.com/vdsc60Gv

**Comment by stmcginnis** [comment]
> Sorry for not being clear. After you delete the extra index, you can restart the docker-compose docker-compose up.

Which of the three entries should be deleted?

> Yes, this works fine for me. I personally use the docker-compose solution for elasticsearch and kibiter. So, I hope it should work fine for you as well.

Well, that's not quite the same then, is it? It appears there is an issue if you clone the repo, go into the docker-compose directory, and run ``docker-compose up``. Those are the instructions in the README, so I'm trying to understand if there is something missing in the documentation, or if something is wrong with the default docker-compose file.

**Comment by vchrombie** [comment]
Sorry @stmcginnis for the inconvenience. I have just pruned all the images and I am trying to perform a fresh installation to see if I can reproduce the issue.

**Comment by stmcginnis** [comment]
Thanks a lot @vchrombie! Would be great to know if it's something I am doing, something with my environment, or something with the code. :)

**Comment by vchrombie** [comment]
Hi @stmcginnis, I have removed all the existing images, volumes and started a fresh installation using the docker-compose solution.

I had to wait for some time for the mordred to start and everything went smoothly.

There were multiple indices created, but the dashboard and everything came out without any problem. :thinking: 
```
$ curl -XGET -k "http://localhost:9200/_alias?pretty"
{
  ".grimoirelab-sigils" : {
    "aliases" : { }
  },
  ".kibana_1" : {
    "aliases" : { }
  },
  ".kibana_2" : {
    "aliases" : {
      ".kibana" : { }
    }
  },
  ...
  ...
}
```

![Screenshot_20200820_030519](https://user-images.githubusercontent.com/25265451/90697798-d7b59a80-e29c-11ea-8244-1526fcde9733.png)

I couldn't reproduce the configuration which you have shared.

> Not sure if this helps, but I kept fetching those aliases during start up. It first created .kibana_1, then .kibana, then .kibana_2. Final result is:
> 
> ```
> $ curl -XGET -k "http://admin:admin@localhost:9200/_alias?pretty"
> {
>   ".kibana" : {
>     "aliases" : { }
>   },
>   ".kibana_1" : {
>     "aliases" : { }
>   },
>   ".kibana_2" : {
>     "aliases" : { }
>   }
> }
> ```

I'll definitely look into this tomorrow if I get time. Meanwhile, @sduenas @zhquan or @mafesan can comment if they find anything useful.

**Comment by vchrombie** [comment]
Hi @stmcginnis, I tried but couldn't reproduce the issue. I couldn't understand what might be causing the issue. Sorry about that.

I would like to ask if you can remove the images, volumes and try using `docker-compose` again.
```
$ docker-compose down
$ docker system prune -a
```

Sorry for the inconvenience caused.

**Comment by stmcginnis** [comment]
Hey @vchrombie, sorry for the long delay getting back to this.

I tried with a fresh Ubuntu Focal instance, and that one worked fine with the latest and greatest system packages and fetch from the repo. Other than initially forgetting to set `sysctl -w vm.max_map_count=262144`, everything did appear to start up normal after adjusting that.

I went back to my original Bionic image and ran the `docker system prune -a`, `git pull --rebase`, and `apt upgrade`. I was then able to start things up, and it all looked good.

So not sure if something had changed in the code in the last couple of weeks, or if some system package update resolved something, but it all appears to work as expected now.

Thanks for your help looking into the failures.
